test/test_listener.cc

```diff
@@ -1,22 +1,26 @@
 #include "test/test_listener.h"
 
 #include "source/common/common/assert.h"
+#include "source/common/runtime/runtime_features.h"
 
 #include "test/test_common/global.h"
 
 namespace Envoy {
 
 void TestListener::OnTestEnd(const ::testing::TestInfo& test_info) {
-  // Check that all singletons have been destroyed.
-  std::string active_singletons = Envoy::Test::Globals::describeActiveSingletons();
-  RELEASE_ASSERT(active_singletons.empty(),
-                 absl::StrCat("FAIL [", test_info.test_suite_name(), ".", test_info.name(),
-                              "]: Active singletons exist. Something is leaking. Consider "
-                              "commenting out this assert and letting the heap checker run:\n",
-                              active_singletons));
-  RELEASE_ASSERT(!Thread::MainThread::isMainThreadActive(),
-                 absl::StrCat("MainThreadLeak: [", test_info.test_suite_name(), ".",
-                              test_info.name(), "] test exited before main thread shut down"));
+  if (validate_singletons_) {
+    // Check that all singletons have been destroyed.
+    std::string active_singletons = Envoy::Test::Globals::describeActiveSingletons();
+    RELEASE_ASSERT(active_singletons.empty(),
+                   absl::StrCat("FAIL [", test_info.test_suite_name(), ".", test_info.name(),
+                                "]: Active singletons exist. Something is leaking. Consider "
+                                "commenting out this assert and letting the heap checker run:\n",
+                                active_singletons));
+    RELEASE_ASSERT(!Thread::MainThread::isMainThreadActive(),
+                   absl::StrCat("MainThreadLeak: [", test_info.test_suite_name(), ".",
+                                test_info.name(), "] test exited before main thread shut down"));
+  }
+  Runtime::RuntimeFeaturesDefaults::get().restoreDefaults();
```

Comment:
Interesting...
My initial thoughts were that this wouldn't work, because some tests will have a Scope defined as a class member, and override a runtime value in the c'tor, but if CI is green pass I guess that's ok.

Reply:
yeah all we need to do is set the flags back between unit tests